### DISCUSSION TITLE
Lossless conversion Heic to PNG

### DISCUSSION THREAD

**Discussion by RyanzpLee** [ImageMagick/ImageMagick#4759] [Help] [answered]
I'm trying to convert a heic file to a more widely support filetype such as png to display in browsers.

I ran this command `magick file.heic file.png` on macOS and it converted it properly. But when I inspect the pixel colour value with `Digital Colour Meter`, it changed from rgb(0,0,255) to rgb(0,0,254) which I do not want to happen. 

I then tried `magick file.heic -quality 100 file.png` which did not retain the correct pixel colour value rgb(0,0,255). And also `magick file.heic -compress lossless file.png` did not retain the correct pixel colour value rgb(0,0,255)

When I open the heic file in macos preview, and then export it to a png file, there is no change in colour values. How can I produce this with imageMagick please?

Example heic file and converted png file in zip.

[Archive.zip](https://github.com/ImageMagick/ImageMagick/files/7943516/Archive.zip)

ImageMagick Version: ImageMagick 7.1.0-20 Q16-HDRI x86_64 2022-01-22 https://imagemagick.org
Platform: macOS Big Sur Version 11.6.2
Libheif Version: 1.12.0_1

**Comment by fmw42** [comment]
> it changed from rgb(0,0,255) to rgb(0,0,255)

Those are the same values!  Please clarify you question. It is not clear to me.  Why did -quality 100 not work -- in what way?  Also why did -compress not work?

Please post an example heic file and your png results. Also state what version of ImageMagick you are using and on what platform and what version of libheif.

**Reply by RyanzpLee** [reply]
I apologise for the typo! I meant (0,0,254). I've fixed that and added the versions. I cannot upload an example heic file but I've added the produced png file, and clarified why some things are not working

**Reply by fmw42** [reply]
Why can you not upload the heic file? You can always Zip it and then upload? Is it proprietary?

**Reply by RyanzpLee** [reply]
> Why can you not upload the heic file? You can always Zip it and then upload? Is it proprietary?

Ah of course, I didn't think of doing that. I've uploaded the zip now. <3

**Reply by fmw42** [reply]
FYI I have found that using many color picker apps, I often get one or more pixel gray level differences when measuring from the screen.  I generally use a full editor tool such as GraphicConverter on the Mac or Photoshop or GIMP to read pixel values. 

I get 0 difference from your files:

`magick compare -metric rmse test.heic test.png null:`
0 (0)

**Reply by RyanzpLee** [reply]
Hmm that's strange because I only noticed the difference at first when I passed both the images to my app which uses https://github.com/mapbox/pixelmatch to diff the images and it showed up as there being a difference between the two images.

Does the rmse metric have a slight threshold which ignores any differences under a specific amount?

**Reply by fmw42** [reply]
No, -metric rmse has no threshold, unless you add -fuzz XX% to your command.

**Reply by fmw42** [reply]
Perhaps you should check with EXIFTOOL.

**Reply by RyanzpLee** [reply]
> No, -metric rmse has no threshold, unless you add -fuzz XX% to your command.

Ah okay I see. That's strange then, I'll have to further investigate.

> Perhaps you should check with EXIFTOOL.

[Is this the tool you're referring to?](https://exiftool.org/) I will check it out

**Comment by fmw42** [comment, marked as answer]
Yes.  But if you are on a Mac, I would suggest you get GraphicConvert from https://www.lemkesoft.de/en/products/graphicconverter/

**Reply by RyanzpLee** [reply]
Thank you for your help! I compared the heic file and the actual original png file, and realised that the rgb values are actually different. In the original png file the blue is rgb(0,0,255) but in the heic file it is (0,0,254), which means the imagemagick conversion does work as intended!

It just seems that the heic compression alters the colour slightly for whatever reason?

**Comment by PacoH** [comment]
OP did not state which conversion the `test.png` file was but I checked it as suggested by  [fmw42](https://github.com/ImageMagick/ImageMagick/discussions/4759#discussioncomment-2052815). I also got 0 (0). I also converted it myself using `magick file.heic -compress lossless file.png` as OP tried and got the same result. However, when I converted it using the horrible `Preview` app I got 851.148 (0.0129877)! So I don't know how he is getting "no change in colour values." I also used `magick` to convert one of my iPhone HEIC images (I don't know how they defaulted to this format but I changed it in settings back to JPG) and I got 0 (0). This is what I was looking for so I can convert all those HEIC images to JPG.

**Comment by IOrlandoni** [comment]
To use the HEIC to PNG conversion tool, simply navigate to [https://convert.app/heic-to-png/](https://convert.app/heic-to-png/) in your web browser. Once there, you can either drag and drop your HEIC files into the designated area on the page, or click the "Select files" button to choose the files you want to convert from your local storage. You can select multiple files at once if you need to convert more than one.

**Comment by masonduug** [comment]
Here is the best way to convert heic to png on both Mac and Windows PC: [https://discussions.apple.com/thread/255331998?sortBy=rank](https://discussions.apple.com/thread/255331998?sortBy=rank)

It supports heic images from iPhone, Android and digital cameras.

**Reply by fmw42** [reply]
Should be the same for Mac And PC so long as you can install the libheif library. There might be restrictions for PC. You will have to check. See https://github.com/strukturag/libheif

`magick image.heic image.png`
